File: python/tests/test_crash_safety.py
"""Crash-safety tests: bounded spin on a stuck slot.

A peer that dies mid-operation leaves its slot state permanently claimed
(WRITING/READING). These tests simulate such a ghost by poking the
slot-state array directly, then assert push/pop/top fail within the bounded
spin instead of hanging, undo their top reservation (no silent item loss),
and that the stack works again once the slot is repaired.

Runs against whichever backend is loaded: with the C FFI present this
exercises c/src/ffi.c's bounded spin; with ZEROIPC_FFI_LIB=none it
exercises the pure-Python fallback's.
"""

import os
import struct

import numpy as np
import pytest

from zeroipc import Memory, Stack

HEADER_SIZE = 16
SLOT_EMPTY = 0
SLOT_WRITING = 1
SLOT_READY = 2
SLOT_READING = 3


def _align8(n: int) -> int:
    return (n + 7) & ~7


@pytest.fixture
def mem():
    name = f"/test_crash_{os.getpid()}"
    m = Memory(name, 1024 * 1024)
    yield m
    m.close()
    m.unlink()


def _state_offset(mem, name, elem_size, capacity, i):
    entry = mem.table.find(name)
    assert entry is not None
    side_off = HEADER_SIZE + _align8(elem_size * capacity)
    return entry.offset + side_off + i * 4


def _poke_state(mem, off, value):
    struct.pack_into("<I", mem.data, off, value)


def test_stack_crashed_peer_bounded_spin(mem):
    stack = Stack(mem, "s", capacity=8, dtype=np.int32)

    assert stack.push(1)
    assert stack.push(2)

    state1 = _state_offset(mem, "s", 4, 8, 1)

    # Ghost pusher died mid-write: slot 1 stuck in WRITING.
    _poke_state(mem, state1, SLOT_WRITING)
    assert stack.pop() is None  # bounded bail-out, not a hang
    assert stack.size() == 2  # undo restored top; nothing lost

    # Repair; both values drain.
    _poke_state(mem, state1, SLOT_READY)
    assert stack.pop() == 2
    assert stack.pop() == 1

    # Ghost popper holding the slot above top: push fails bounded, top restored.
    assert stack.push(1)
    _poke_state(mem, state1, SLOT_READING)
    assert not stack.push(2)
    assert stack.size() == 1

    _poke_state(mem, state1, SLOT_EMPTY)
    assert stack.push(2)
    assert stack.size() == 2


def test_stack_top_bails_out_on_stuck_slot(mem):
    stack = Stack(mem, "t", capacity=8, dtype=np.int32)
    assert stack.push(42)

    state0 = _state_offset(mem, "t", 4, 8, 0)
    _poke_state(mem, state0, SLOT_WRITING)
    assert stack.top() is None

    _poke_state(mem, state0, SLOT_READY)
    assert stack.top() == 42

File: cpp/tests/test_crash_safety.cpp
#include <gtest/gtest.h>
#include <zeroipc/memory.h>
#include <zeroipc/stack.h>
#include <atomic>

using namespace zeroipc;

// A peer that crashes mid-operation leaves its slot state permanently
// claimed (WRITING/READING). These tests simulate such a ghost by poking the
// slot-state array directly, then assert that push/pop/top fail within their
// bounded spin instead of hanging, that they undo their top reservation (no
// silent item loss), and that the stack works again once the slot is
// repaired.

class CrashSafetyTest : public ::testing::Test {
protected:
    void SetUp() override { Memory::unlink("/test_crash_safety"); }
    void TearDown() override { Memory::unlink("/test_crash_safety"); }

    // Pointer to slot i's state, computed from the spec layout formula.
    static std::atomic<uint32_t>* state_ptr(Memory& mem, const char* name,
                                            size_t elem_size, size_t cap,
                                            size_t i) {
        size_t offset = 0, size = 0;
        if (!mem.find(name, offset, size)) return nullptr;
        size_t side_off = 16 + align_up(elem_size * cap, 8);
        return reinterpret_cast<std::atomic<uint32_t>*>(
            static_cast<char*>(mem.base()) + offset + side_off + i * 4);
    }
};

TEST_F(CrashSafetyTest, PopBailsOutWhenSlotStuckInWriting) {
    Memory mem("/test_crash_safety", 1024 * 1024);
    Stack<int> s(mem, "s", 8);

    ASSERT_TRUE(s.push(1));
    ASSERT_TRUE(s.push(2));

    auto* st = state_ptr(mem, "s", sizeof(int), 8, 1);
    ASSERT_NE(st, nullptr);

    // Ghost: a pusher died mid-write, slot 1 stuck in WRITING.
    st->store(Stack<int>::SLOT_WRITING);

    // Pop must fail bounded (not hang) and restore top: nothing lost.
    EXPECT_FALSE(s.pop().has_value());
    EXPECT_EQ(s.size(), 2u);

    // Repair the slot; the stack works again and no item was dropped.
    st->store(Stack<int>::SLOT_READY);
    auto v = s.pop();
    ASSERT_TRUE(v.has_value());
    EXPECT_EQ(*v, 2);
    v = s.pop();
    ASSERT_TRUE(v.has_value());
    EXPECT_EQ(*v, 1);
}

TEST_F(CrashSafetyTest, PushBailsOutWhenSlotStuckAboveTop) {
    Memory mem("/test_crash_safety", 1024 * 1024);
    Stack<int> s(mem, "s", 8);

    ASSERT_TRUE(s.push(1));
    ASSERT_TRUE(s.push(2));

    auto* st = state_ptr(mem, "s", sizeof(int), 8, 2);
    ASSERT_NE(st, nullptr);

    // Ghost: a popper died holding slot 2 (above the current top).
    st->store(Stack<int>::SLOT_READING);

    // Push must fail bounded and restore top.
    EXPECT_FALSE(s.push(3));
    EXPECT_EQ(s.size(), 2u);

    // Repair; push works again.
    st->store(Stack<int>::SLOT_EMPTY);
    EXPECT_TRUE(s.push(3));
    auto v = s.pop();
    ASSERT_TRUE(v.has_value());
    EXPECT_EQ(*v, 3);
}

TEST_F(CrashSafetyTest, TopBailsOutWhenSlotStuck) {
    Memory mem("/test_crash_safety", 1024 * 1024);
    Stack<int> s(mem, "s", 8);

    ASSERT_TRUE(s.push(42));

    auto* st = state_ptr(mem, "s", sizeof(int), 8, 0);
    ASSERT_NE(st, nullptr);

    st->store(Stack<int>::SLOT_WRITING);
    EXPECT_FALSE(s.top().has_value());

    st->store(Stack<int>::SLOT_READY);
    auto v = s.top();
    ASSERT_TRUE(v.has_value());
    EXPECT_EQ(*v, 42);
}
